src/channel.c

```diff
@@ -2111,6 +2111,77 @@ channel_fill(js_read_T *reader)
     return TRUE;
 }
 
+/*
+ * Process the HTTP header in a Language Server Protocol (LSP) message.
+ *
+ * The message format is described in the LSP specification:
+ * https://microsoft.github.io/language-server-protocol/specification
+ *
+ * It has the following two fields:
+ *
+ *	Content-Length: ...
+ *	Content-Type: application/vscode-jsonrpc; charset=utf-8
+ *
+ * Each field ends with "\r\n". The header ends with an additional "\r\n".
+ *
+ * Returns OK if a valid header is received and FAIL if some fields in the
+ * header are not correct. Returns MAYBE if a partial header is received and
+ * need to wait for more data to arrive.
+ */
+    static int
+channel_process_lsp_http_hdr(js_read_T *reader)
+{
+    char_u *line_start;
+    char_u *p;
+    int	    hdr_len;
+    int	    payload_len = -1;
+    int	    jsbuf_len;
+
+    // We find the end once, to avoid calling strlen() many times.
+    jsbuf_len = STRLEN(reader->js_buf);
+    reader->js_end = reader->js_buf + jsbuf_len;
+
+    p = reader->js_buf;
+
+    // Process each line in the header till an empty line is read (header
+    // separator).
+    while (TRUE)
+    {
+	line_start = p;
+	while (*p != NUL && *p != '\n')
+	    p++;
+	if (*p == NUL)			// partial header
+	    return MAYBE;
+	p++;
+
+	// process the content length field (if present)
+	if ((p - line_start > 16)
+		&& STRNICMP(line_start, "Content-Length: ", 16) == 0)
+	    payload_len = atoi((char *)line_start + 16);
+
+	if ((line_start + 2) == p && line_start[0] == '\r' &&
```

Comment:
Minor nit, make this consistent with the check above.

```suggestion
	if ((p - line_start) == 2 && line_start[0] == '\r' &&
```